rotkehlchen/chain/evm/decoding/compound/v3/balances.py

```diff
@@ -0,0 +1,137 @@
+from rotkehlchen.chain.ethereum.utils import token_normalized_value_decimals
+from rotkehlchen.constants import ZERO
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.errors.asset import UnknownAsset, WrongAssetType
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTokenKind
+
+from .constants import CPT_COMPOUND_V3
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.contracts import EvmContract
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Compoundv3Balances(ProtocolWithBalance):
+    def __init__(self, database: 'DBHandler', evm_inquirer: 'EvmNodeInquirer'):
+        super().__init__(
+            database=database,
+            evm_inquirer=evm_inquirer,
+            counterparty=CPT_COMPOUND_V3,
+            deposit_event_types={(HistoryEventType.DEPOSIT, HistoryEventSubType.DEPOSIT_ASSET)},
+        )
+
+    def _extract_unique_borrowed_tokens(self) -> tuple[dict['EvmToken', set['ChecksumEvmAddress']], dict['ChecksumEvmAddress', 'EvmToken']]:  # noqa: E501
+        """
+        Fetch borrow unique events from the userDB. Since a user can increase or decrease the same
+        liability, we remove the duplicates to reduced the amount of queries. Returns a dict of
+        compound token's address -> set of addresses which borrowed the underlying token, and a
+        dict of compound token's address -> its underlying token.
+        """
+        unique_borrows: dict['EvmToken', set['ChecksumEvmAddress']] = defaultdict(set)
+        underlying_token: dict['ChecksumEvmAddress', 'EvmToken'] = {}
+        for address, events in self.addresses_with_activity(
+            products=None,
+            event_types={(HistoryEventType.RECEIVE, HistoryEventSubType.GENERATE_DEBT)},
+        ).items():
+            for event in events:
+                if event.address is not None:
+                    try:
+                        _compound_token = EvmToken(evm_address_to_identifier(
+                            address=event.address,
+                            chain_id=self.evm_inquirer.chain_id,
+                            token_type=EvmTokenKind.ERC20,
+                        ))
+                        if len(_compound_token.underlying_tokens) == 0:
+                            log.error(f'No underlying token found for {_compound_token!s}.')
+                            continue
+
+                        _underlying_token = EvmToken(evm_address_to_identifier(
+                            address=_compound_token.underlying_tokens[0].address,
```

Comment:
Why the `_` in front of the variable? I believe it doesn't follow our style guide